### PR TITLE
wip(frontend): add spreadsheet editor widget (AYC-153)

### PR DESCRIPTION
**Pull request by pavellnk**
- react-datasheet-grid based editable grid (React 19 peer-verified),
  lazy-load ready: the lib CSS is imported inside the widget
- Lenient content parser falls back to an empty sheet instead of
  throwing; grid uses positional column keys (c0..cN) so renames and
  duplicate labels stay unambiguous
- Numeric coercion on save only ('' -> null, plain decimals -> numbers,
  identifier-like strings such as '007' stay strings)
- Column add/rename/delete as pure tested functions; delete confirms
  only when the column holds data
- Explicit Save disabled when clean; historical versions render
  read-only; dark mode via --dsg-* variable overrides

---

> [!NOTE]
> **Medium Risk**
> Large new user-facing editor with artifact content serialization and version/history edge cases; mitigated by tests and defensive parsing, but save/export integration and data loss paths (column delete, invalid overwrite) warrant careful review.
> 
> **Overview**
> Introduces a new **`spreadsheet-editor`** widget for spreadsheet artifacts: editable grid via **`@revolist/react-datagrid`**, explicit save, XLSX/CSV export hooks, and shared **`VersionHistory`** integration.
> 
> **Content model:** `spreadsheet-v1` JSON with a **lenient parser** (corrupt content → empty sheet + warning, no crash). Grid state uses **positional column keys** (`c0`…`cN`) so renames and duplicate headers stay safe; **numeric coercion runs on save only** (e.g. `007` stays a string).
> 
> **Editing UX:** Toolbar for batch add rows (capped at 5000 rows / 100 columns), column manager with rename, delete (confirm when data exists), and drag reorder (`@dnd-kit`). **Read-only** when browsing older versions; dirty state blocks history navigation and gates revert/select. Close flow prompts to save, discard, or keep editing (including empty dirty sheets).
> 
> **`VersionHistory` changes:** Adds **`selectedVersionNumber`** so the highlighted row can differ from the server “current” version; **revert is no longer nested** inside the version select button (accessibility fix + test).
> 
> EN/DE **`artifacts.spreadsheet`** strings and widget **`SUMMARY.md`** updated; broad unit test coverage for parsers, grid ops, editor state, grid source stability, and close/save behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 86a001475fc1813968c6cf9ecaefaeebc82c2543. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by pavellnk** [comment]
* **#1289** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1289?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1061** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1061?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1055** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1055?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1054** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1054?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1054?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1052** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1052?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1053](https://github.com/ayunis-core/ayunis-core/pull/1053) <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1053?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1051** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1051?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1050** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1050?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1049** <a href="https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1049?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: ayunis-core-frontend/src/widgets/spreadsheet-editor/model/useSpreadsheetEditorState.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  const edit = (updater: (state: GridState) => GridState) => {
+    setGridState(updater);
+    setIsDirty(true);
+  };
+
+  return {
+    displayedGridState,
+    isDirty,
+    isValid: loaded.isValid,
```
[the remark this answers is not in the document]
Comment (pavellnk) [reply ayunis-core-frontend/src/widgets/spreadsheet-editor/model/useSpreadsheetEditorState.ts]: Fixed in the latest push: isValid is now state and is recomputed in the version-reload effect, so the banner clears once a valid version loads.

File: ayunis-core-frontend/src/widgets/spreadsheet-editor/ui/SpreadsheetEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}: SpreadsheetEditorProps) {
+  const { t } = useTranslation('artifacts');
+  const { confirm } = useConfirmation();
+  const editor = useSpreadsheetEditorState(artifact);
+
+  const handleExport = (format: SpreadsheetExportFormat) => {
+    onExport(
+      format,
+      editor.isDirty ? editor.getSerializedContent() : undefined,
+    );
```
[the remark this answers is not in the document]
Comment (pavellnk) [reply ayunis-core-frontend/src/widgets/spreadsheet-editor/ui/SpreadsheetEditor.tsx:56]: Fixed in the latest push: export now serializes the displayed grid whenever the user is viewing history or has unsaved edits, so downloads always match the screen.

File: ayunis-core-frontend/src/widgets/spreadsheet-editor/ui/SpreadsheetEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          {t('spreadsheet.invalidContent')}
+        </div>
+      )}
+
+      <div className="flex-1 overflow-y-auto p-2">
+        <SpreadsheetGrid
+          columns={editor.displayedGridState.columns}
+          rows={editor.displayedGridState.rows}
+          onRowsChange={editor.setRows}
+          readOnly={editor.isViewingHistory}
```
[the remark this answers is not in the document]
Comment (pavellnk) [reply ayunis-core-frontend/src/widgets/spreadsheet-editor/ui/SpreadsheetEditor.tsx:144]: Fixed in the latest push: the hook's edit() ignores all mutations while viewing history, so no grid event can dirty or replace the editable state from that mode.

File: ayunis-core-frontend/src/widgets/spreadsheet-editor/ui/SpreadsheetEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        />
+      </div>
+
+      {artifact.versions && artifact.versions.length > 0 && (
+        <VersionHistory
+          versions={artifact.versions}
+          currentVersionNumber={editor.displayedVersionNumber}
+          onSelect={editor.selectVersion}
+          onRevert={onRevert}
+        />
```
[the remark this answers is not in the document]
Comment (pavellnk) [reply ayunis-core-frontend/src/widgets/spreadsheet-editor/ui/SpreadsheetEditor.tsx:155]: Fixed in the latest push: VersionHistory now receives the artifact's true currentVersionNumber (label + Revert visibility) plus a new selectedVersionNumber prop for the row highlight, so a previewed version keeps its Revert action.

File: ayunis-core-frontend/src/widgets/spreadsheet-editor/model/spreadsheet-content.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  };
+}
+
+export function serializeSpreadsheetContent(data: SpreadsheetContent): string {
+  return JSON.stringify({
+    format: SPREADSHEET_CONTENT_FORMAT,
+    columns: data.columns,
+    rows: data.rows,
+  });
+}
```
[the remark this answers is not in the document]
Comment (pavellnk) [reply ayunis-core-frontend/src/widgets/spreadsheet-editor/model/spreadsheet-content.ts]: Fixed in the latest push: Save is disabled when the sheet has no columns, so the backend's non-empty-columns rule can no longer be hit from a valid UI action.

**Review by devbydaniel — APPROVED** [review]

**Review by sirolad — APPROVED** [review]

**Comment by devbydaniel** [comment]
### Merge activity

* **Aug 6, 7:42 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1054).
* **Aug 6, 7:49 PM UTC**: [Graphite](https://app.graphite.com/github/pr/ayunis-core/ayunis-core/1054) couldn't merge this PR because it had merge conflicts.
